### PR TITLE
Add Storage Extension Support

### PR DESCRIPTION
**Pull request by MicroDev1**
Adds ability to extend `CIRCUITPY` storage by utilizing the next update partition on boards with `dualbank` support.

The `ota_x` and `user_fs` partitions are presented as a continuous flash storage block with `user_fs` always forming the first part and the second part is one of the `ota_x` next update partition. These partition may not be contiguous in the `partition-table`.

Couple limitations as re-sizing filesystem without erasing isn't supported currently:
1. Filesystem needs to be erased when re-sizing.
2. If both `ota_x` app partitions need to contain bootable image, storage extension has to be turned off.

Note: TinyUF2 will need to be updated so it flashes the next update partition instead of defaulting to `ota_0` in order to avoid data corruption when storage is extended.

### CODE REVIEW

**Comment by bill88t** [comment]
Is this ready for testing?
Or do we first have to wait for tinyuf2?

**Comment by MicroDev1** [comment]
@bill88t You can start with testing.

**Comment by bill88t** [comment]
Manually built f44288d and flashed on waveshare_esp32s2_pico as a uf2.

Had to look in the source to find it. `storage.erase_filesystem(extended=True)`
This needs to be documented.

Attempting a wget test.

image

Works just fine.
Let's try some more stuff.

image

It works if we go back to the first partition and to the 2nd again.

**Comment by bill88t** [comment]
If `extended` is not specified it assumes True, extending the storage. I did test that.
`extended=False` also works as expected.

File: shared-bindings/storage/__init__.c [hunk 1]
```diff
@@ -150,7 +150,7 @@ STATIC mp_obj_t storage_getmount(const mp_obj_t mnt_in) {
 }
 MP_DEFINE_CONST_FUN_OBJ_1(storage_getmount_obj, storage_getmount);
 
-//| def erase_filesystem() -> None:
+//| def erase_filesystem(extended: Optional[bool] = True) -> None:
```
Comment (jepler) [inline shared-bindings/storage/__init__.c:153]: I think we need to discuss what the default value of this and dualbank.switch(both_bootable=) should be. Also, what the action in the case of corrupt CIRCUITPY is, and whether it should match the default of an explicit `erase_filesystem()` with its default value. If this has already been discussed, a pointer to the discussion would be helpful.

File: ports/espressif/supervisor/internal_flash.c [hunk 2]
```diff
@@ -32,47 +33,114 @@
...
+    } else {
+        esp_partition_erase_range(partition, offset, size_total);
+        esp_partition_write(partition, offset, _cache, size_total);
+    }
+}
+
+#if CIRCUITPY_STORAGE_EXTEND
+STATIC void multi_partition_rw(uint8_t *data,
+    const uint32_t offset, const uint32_t size_total, const bool op) {
+    if (offset > _partition[0]->size) {
```
Comment (jepler) [inline ports/espressif/supervisor/internal_flash.c:112]: A suggestion: the `#if CIRCUITPY_STORAGE_EXTENDS`s below I think could be eliminated if you had an else branch here that did something like `#define multi_partition_rw(data, offset, size_total, op) single_partition_rw(_partition[0], data, offset, size_total, op)`, or a function to do the same. However, if you don't think that's a net improvement in readability that's fine as well.

File: shared-bindings/storage/__init__.c (hunk 1 above)
Comment (MicroDev1) [reply shared-bindings/storage/__init__.c:153]: I chose to make storage extended by default as I think more people would want an extended storage over ota updates.

Comment (MicroDev1) [reply shared-bindings/storage/__init__.c:153]: Now that you pointed it out, I realized that this leaves the possibility where the user wants both partitions to contain bootable image but corrupts the CIRCUITPY, calls just `erase_filesystem()` and ends up having the other image wiped.

Comment (MicroDev1) [reply shared-bindings/storage/__init__.c]: Using the following API now:
```py
def erase_filesystem(extended: Optional[bool] = None) -> None:
    """Erase and re-create the ``CIRCUITPY`` filesystem.

    :param bool extended: On boards that support ``CIRCUITPY_DUALBANK``
        and ``CIRCUITPY_STORAGE_EXTEND``, the ``CIRCUITPY`` storage can be
        extended by setting this to `True`. If this isn't provided or set to
        `None` (default), existing configuration will be used."""
```
Using `dualbank` when storage is extended will raise an error.
```c
STATIC void raise_error_if_storage_extended(void) {
    if (supervisor_flash_get_extended()) {
        mp_raise_RuntimeError(translate("storage is extended"));
    }
}
```

File: ports/espressif/common-hal/dualbank/__init__.c [hunk 3]
```diff
@@ -32,6 +32,8 @@
 #include "esp_log.h"
 #include "esp_ota_ops.h"
 
+#include "supervisor/flash.h"
+
```
Comment (jepler) [inline ports/espressif/common-hal/dualbank/__init__.c:36]: Is this addition needed?

File: shared-bindings/storage/__init__.c [hunk 4]
```diff
@@ -160,16 +161,36 @@ MP_DEFINE_CONST_FUN_OBJ_1(storage_getmount_obj, storage_getmount);
 //|     This function can be called from the REPL when ``CIRCUITPY``
 //|     has become corrupted.
 //|
+//|     :param bool extended: On boards that support ``CIRCUITPY_DUALBANK``
+//|         and ``CIRCUITPY_STORAGE_EXTEND``, the ``CIRCUITPY`` storage can be
```
Comment (dhalbert) [inline shared-bindings/storage/__init__.c:165]:
```suggestion
//|     :param bool extended: On boards that support ``dualbank``
//|         and the ``extended`` parameter, the ``CIRCUITPY`` storage can be
```

**Review by jepler — COMMENTED** [review]
I appreciate you re-working how the storage.erase_filesystem call will work.
I have one minor nit, though.

File: shared-bindings/storage/__init__.c [hunk 5]
```diff
@@ -160,16 +161,36 @@ MP_DEFINE_CONST_FUN_OBJ_1(storage_getmount_obj, storage_getmount);
...
+//|         extended by setting this to `True`. If this isn't provided or set to
+//|         `None` (default), exisiting configuration will be used.
```
Comment (dhalbert) [inline shared-bindings/storage/__init__.c:167]:
```suggestion
//|         `None` (default), the existing configuration will be used.
//|         If the ``extend`` parameter is not supported but is set to other than ``None``,
//|         a `ValueError` will be raised.
```

File: shared-bindings/storage/__init__.c [hunk 6]
```diff
@@ -160,16 +161,36 @@ MP_DEFINE_CONST_FUN_OBJ_1(storage_getmount_obj, storage_getmount);
...
+
+    #if CIRCUITPY_STORAGE_EXTEND
+    // parse args
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+    // set extended
+    bool extended = (args[ARG_extended].u_obj == mp_const_none) ? supervisor_flash_get_extended() : mp_obj_is_true(args[ARG_extended].u_obj);
+    common_hal_storage_erase_filesystem(extended);
+    #else
+    common_hal_storage_erase_filesystem(false);
```
Comment (dhalbert) [inline shared-bindings/storage/__init__.c:188]: I think it would be good to check for the argument being other than ``None`` if CIRCUITPY_STORAGE_EXTEND is not 1, and to throw a ValueError in that case, to make it clear it isn't supported.

**Review by dhalbert — DISMISSED** [review]
I like this API -- thanks!

File: shared-bindings/storage/__init__.c (hunk 6 above)
Comment (jepler) [reply shared-bindings/storage/__init__.c:188]: I was tempted to make the new parameter not even be in the Python API if it's not supported, but maybe that's too heretical. It'd be nice not to give up those bytes of flash.

Comment (dhalbert) [reply shared-bindings/storage/__init__.c:188]: I see what you are saying about flash size, yes, it's another 24 bytes or whatever.

Comment (dhalbert) [reply shared-bindings/storage/__init__.c:190]: I made one minor change here to use NotImplementedError instead of adding a new message.

File: shared-bindings/storage/__init__.c [hunk 7]
```diff
@@ -160,16 +161,38 @@ MP_DEFINE_CONST_FUN_OBJ_1(storage_getmount_obj, storage_getmount);
...
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    #if CIRCUITPY_STORAGE_EXTEND
+    bool extended = (args[ARG_extended].u_obj == mp_const_none) ? supervisor_flash_get_extended() : mp_obj_is_true(args[ARG_extended].u_obj);
+    common_hal_storage_erase_filesystem(extended);
+    #else
+    if (args[ARG_extended].u_obj != mp_const_none) {
+        mp_raise_ValueError_varg(translate("%q isn't supported"), MP_QSTR_extended);
```
Comment (dhalbert) [inline shared-bindings/storage/__init__.c:188]: Use NotImplementedError instead, and then we can also avoid adding a new string. (You'll need to `make translate again).
```suggestion
        mp_raise_NotImplementedError_varg(translate("%q"), MP_QSTR_extended);
```

**Review by dhalbert — CHANGES_REQUESTED** [review]
Just one thing!

File: shared-bindings/storage/__init__.c (hunk 7 above)
Comment (MicroDev1) [reply shared-bindings/storage/__init__.c:188]: How about the following:
```c
if (args[ARG_extended].u_obj != mp_const_none && mp_obj_is_true(args[ARG_extended].u_obj)) {
    mp_raise_NotImplementedError(translate("%q=%q"), MP_QSTR_extended, MP_QSTR_True);
}
```

Comment (dhalbert) [reply shared-bindings/storage/__init__.c:188]: that's fine too

Comment (dhalbert) [reply shared-bindings/storage/__init__.c:188]: and even clearer

**Review by dhalbert — APPROVED** [review]
Glad to get this in -- it will be in beta.2, which is imminent.
